commit af9597adc2f1e3609c67c9792a2469bb64e43ae9
Author: Shuai Xue <xueshuai@linux.alibaba.com>
Date:   Fri Dec 8 10:56:51 2023 +0800

    drivers/perf: add DesignWare PCIe PMU driver
    
    This commit adds the PCIe Performance Monitoring Unit (PMU) driver support
    for T-Head Yitian SoC chip. Yitian is based on the Synopsys PCI Express
    Core controller IP which provides statistics feature. The PMU is a PCIe
    configuration space register block provided by each PCIe Root Port in a
    Vendor-Specific Extended Capability named RAS D.E.S (Debug, Error
    injection, and Statistics).
    
    To facilitate collection of statistics the controller provides the
    following two features for each Root Port:
    
    - one 64-bit counter for Time Based Analysis (RX/TX data throughput and
      time spent in each low-power LTSSM state) and
    - one 32-bit counter for Event Counting (error and non-error events for
      a specified lane)
    
    Note: There is no interrupt for counter overflow.
    
    This driver adds PMU devices for each PCIe Root Port. And the PMU device is
    named based the BDF of Root Port. For example,
    
        30:03.0 PCI bridge: Device 1ded:8000 (rev 01)
    
    the PMU device name for this Root Port is dwc_rootport_3018.
    
    Example usage of counting PCIe RX TLP data payload (Units of bytes)::
    
        $# perf stat -a -e dwc_rootport_3018/Rx_PCIe_TLP_Data_Payload/
    
    average RX bandwidth can be calculated like this:
    
        PCIe TX Bandwidth = Rx_PCIe_TLP_Data_Payload / Measure_Time_Window
    
    Signed-off-by: Shuai Xue <xueshuai@linux.alibaba.com>
    Reviewed-by: Baolin Wang <baolin.wang@linux.alibaba.com>
    Reviewed-by: Jonathan Cameron <Jonathan.Cameron@huawei.com>
    Reviewed-by: Yicong Yang <yangyicong@hisilicon.com>
    Reviewed-and-tested-by: Ilkka Koskinen <ilkka@os.amperecomputing.com>
    Link: https://lore.kernel.org/r/20231208025652.87192-5-xueshuai@linux.alibaba.com
    [will: Fix sparse error due to use of uninitialised 'vsec' symbol in
     dwc_pcie_match_des_cap()]
    Signed-off-by: Will Deacon <will@kernel.org>

diff --git a/drivers/perf/Kconfig b/drivers/perf/Kconfig
index 273d67ecf6d2..ec6e0d9194a1 100644
--- a/drivers/perf/Kconfig
+++ b/drivers/perf/Kconfig
@@ -217,6 +217,13 @@ config MARVELL_CN10K_DDR_PMU
 	  Enable perf support for Marvell DDR Performance monitoring
 	  event on CN10K platform.
 
+config DWC_PCIE_PMU
+	tristate "Synopsys DesignWare PCIe PMU"
+	depends on PCI
+	help
+	  Enable perf support for Synopsys DesignWare PCIe PMU Performance
+	  monitoring event on platform including the Alibaba Yitian 710.
+
 source "drivers/perf/arm_cspmu/Kconfig"
 
 source "drivers/perf/amlogic/Kconfig"